File: leetcode/implement-trie-prefix-tree/implement-trie-prefix-tree.rb
# LeetCode #208 - Implement Trie (Prefix Tree)

class Trie
  def initialize()
    @root = {}
  end

=begin
  :type word: String
  :rtype: Void
=end
  def insert(word)
    node = @root
    word.each_char { |ch| node = node[ch] ||= {} }
    node[:word] = true
    nil
  end

=begin
  :type word: String
  :rtype: Boolean
=end
  def search(word)
    !!traverse(word)&.dig(:word)
  end

=begin
  :type prefix: String
  :rtype: Boolean
=end
  def starts_with(prefix)
    !!traverse(prefix)
  end

  private

  def traverse(prefix)
    node = @root
    prefix.each_char { |ch| break unless (node = node[ch]) }
    node
  end
end

# Your Trie object will be instantiated and called as such:
# obj = Trie.new()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.starts_with(prefix)

File: leetcode/implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
// LeetCode #208 - Implement Trie (Prefix Tree)

class Trie {
public:
    Trie() noexcept
    {
        make_node(); // Create the root node.
    }

    void insert(string_view word) noexcept
    {
        auto node = root();

        for (const auto ch : word) {
            if (!ch) abort(); // Null characters not allowed in words.
            auto& child = (*node)[ch];
            if (!child) child = make_node();
            node = child;
        }

        (*node)['\0'] = nullptr;
    }

    [[nodiscard]] bool search(string_view word) const noexcept
    {
        const auto node = traverse(word);
        return node && node->count('\0');
    }

    [[nodiscard]] bool startsWith(string_view prefix) const noexcept
    {
        return traverse(prefix);
    }

private:
    class Node : public unordered_map<char, Node*> { };

    Node* make_node() noexcept
    {
        return &nodes_.emplace_back();
    }

    [[nodiscard]] const Node* root() const noexcept
    {
        return &nodes_.front();
    }

    [[nodiscard]] Node* root() noexcept
    {
        return &nodes_.front();
    }

    [[nodiscard]] const Node* traverse(string_view prefix) const noexcept
    {
        auto node = root();

        for (const auto ch : prefix) {
            if (!ch) abort(); // Null characters not allowed in words.
            const auto it = node->find(ch);
            if (it == end(*node)) return nullptr;
            node = it->second;
        }

        return node;
    }

    deque<Node> nodes_;
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
